CMakeLists.txt

```diff
@@ -521,6 +521,19 @@ endif()
 
 include(CMakePackageConfigHelpers)
 
+if(DEFINED OPENTELEMETRY_BUILD_DLL)
+  if(NOT MSVC)
+    message(WARNING "Build DLL is supposed to work with MSVC!")
+  endif()
+  if(WITH_STL)
+    message(
+      STATUS_FATAL
+      "Build DLL with C++ STL is not recommended because ABI compatibility could be broken!"
```

Comment:
This error message is actually confusing. Should we just say "Build DLL with C++ STL is not supported!", in case we plan to not support it due to technical challenges.

Reply:
> All,
> 
> Following comments in each comment thread is getting hard, so summarizing here the most critical points found so far.
> 
> I have more comments on less critical parts, saving them for later to not divert the discussion.
> 
> ## Technical solution
> First, I am not convinced the following code works.
> 
> ```
> //
> // Partial specialization of default_delete used by unique_ptr or shared_ptr.
> // This makes the delete of the type in unique_ptr/shared_ptr happening in the
> // DLL in which it is allocated.
> //
> template <>
> struct OPENTELEMETRY_EXPORT default_delete<OPENTELEMETRY_NAMESPACE::sdk::trace::SpanExporter>
> {
> public:
>   void operator()(OPENTELEMETRY_NAMESPACE::sdk::trace::SpanExporter *span_exporter)
>   {
>     delete span_exporter;
>   }
> };
> ```
> 
> 
> default_delete is used for std::unique_ptr It is _not_ used by std::shared_ptr, as found recently looking at cppreference.
> 
> std::shared_ptr uses a _delete expression_, as in `delete ptr` instead.
> 
> In my understanding, `std::shared_ptr<T>` will never call default_delete, making the fix incomplete.
> 
> Changing nostd::shared_ptr is not going to address std::shared_ptr
> 
> 
> this is an inline template, exposed in header files, and compiled with the application code compiling unit.
> 
> Regardless of what OPENTELEMETRY_EXPORT expands to, this is still inlined code that will invoke `delete span_exporter`, from the application binary, and not from the opentelemetry dll.
> 
> I don't understand what is achieved here, compared to the default std::unique_ptr code, which invokes default_delete, also calling `delete ptr` from the application space.
> 
> I would understand the OPENTELEMETRY_EXPORT part if this was not an inline definition but only a declaration, with the definition in the DLL, but this is not the case.
> 
> I admit I am not familiar with DLL IMPORT/EXPORT, so it could be I missed something here.
> 
> In that case, please explain, with a lot of details, why and HOW this code works.
> 
> Also, a description of the technical issue to resolve would help here: compared to linux shared libraries, what is it with windows DLL that introduce additional or different constraints ?
> 
> ## Verification
> The fact that unit tests runs in CI is inconclusive, and does not prove the fix works.
> 
> Do the tests actually expose the broken behavior in the first place, that is, destroying memory from the application space (the unit test) when removing the last reference of a unique_ptr or shared_ptr that was allocated in the opentelemetry space ?
> 
> If the last reference to an object is removed from inside the opentelemetry DLL (see CleanupTracer() in examples), the original bug is not exposed, and the code change can not be tested.
> 
> Take this test for example:
> 
> ```
> TEST(OStreamSpanExporter, Shutdown)
> {
>   auto exporter  = std::unique_ptr<trace_sdk::SpanExporter>(new exportertrace::OStreamSpanExporter);
>   auto processor = std::shared_ptr<trace_sdk::SpanProcessor>(
>       new trace_sdk::SimpleSpanProcessor(std::move(exporter)));
> 
>   auto recordable = processor->MakeRecordable();
>   recordable->SetName("Test Span");
> 
>   // Capture the output of cout
>   const auto captured = WithOStreamCapture(std::cout, [&]() {
>     EXPECT_TRUE(processor->Shutdown());
>     processor->OnEnd(std::move(recordable));
>   });
>   std::string err_message =
>       "[Ostream Trace Exporter] Exporting 1 span(s) failed, exporter is shutdown";
>   EXPECT_TRUE(captured.find(err_message) != std::string::npos);
> }
> ```
> 
> `exporter` and `processor` are allocated locally, from the unit test, and not allocated from the opentelemetry-cpp library.
> 
> In fact, this test links with opentelemetry_exporter_ostream_span directly, not with the DLL at all.
> 
> The only place that links with the DLL is examples, and examples are not executed in CI.
> 
> ## Technical risk
> Currently, the build can use WITH_STL=ON or WITH_STL=OFF
> 
> Ultimately, the DDL should exist for both builds.
> 
> Given how unique_ptr and shared_ptr are involved, starting to implement a solution for nostd only is a major red flag in my opinion: once this works for nostd, there is no guarantee that the solution can be transposed to std, especially is code changes that deviate from the std api are implemented in nostd.
> 
> It will be less risk to implement a DDL for WITH_STL=ON first, and then adapt the code to support WITH_STL=OFF second.
> 
> ## Deployment
> As discussed previously with packaging in general, I don't see how a single DDL can be shipped for the SDK and all exporters.
> 
> We don't need to cut the SDK into smaller parts (per signal), so a single DDL for the entire SDK should be ok.
> 
> For exporters, each exporter should have its own DDL in my opinion.
> 
> If somehow a single DDL is desirable, please describe how is it supposed to be used, I just don't get it.

1. Yes, `std::shared_ptr<>` issue is open, but as we don't support `WITH_STL` with DLL build, this is fine.

I thought about fixing `std::shared_ptr<>`. For `WITH_STL`, we can still construct it with a second explicit parameter of deleter for `std::shared_ptr<>`, but the issue is the user's code can drive either `TracerProvider` and `LoggerProvider` and implements their own GetLogger/GetTracer which returns `std::shared_ptr<>` without deleter. So instead of fixing that for SDK but leave holes to the user, I'd prefer we just disable the scenario from the beginning.

Also, not support `WITH_STL` with DLL is mainly not for the reason that it is hard for us/maintainer to implement it, the rationale is it is risky for user to bring their own STL to the binary which will be passed to our DLL, because that STL could be in different version our the one compiled into DLL, then causing inconsistency between binary layout.

2.  `OPENTELEMETRY_EXPORT` is expanded to `__declspec(dllimport)` in the application code, so the referenced symbol like the `operator()` function will be marked as import (symbol name changed after decoratoin) which only resolves to the one marked as `__declspec(dllexport)` in the DLL. It will not resolve to the local definition and could get unresolved link error if it is not exported in DLL.

## Verification
Agree it is lacking testing for now. Not all tests can be converted to using the DLL, like the one you listed as `OStreamSpanExporter::Shutdown` which tests the internal SDK behavior, and should be considered as external user/component of the DLL. For example, SDK user is not expected to new `SpanExporter` or its derived classes directly, the factory class like `OStreamSpanExporterFactory` should be used instead. So the examples are in perfect position of doing this migration which demos the real user scenario. But yes, examples are not tests. let me create some DLL specific tests which only relies on the API and exposed SDK functions. Let me know if there is any concern on this.

## Technical risk
As mentioned in 1. above, don't support WITH_STL in DLL build gives us long term benefit which less fragile on ABI compatibility. But I am also open to hear any customer concern like WITH_STL is required even if they call into our DLL. And anyway, the DLL build will be in preview for some time and we can still make change to it after this PR.

## Deployment
For separating SDK and exporters into different DLLs, we discussed this internally. But this requires us to expose more functions in SDK for exporter use, and dealing with more potential inconsistent memory alloc/free between SDK and exporters. Further more, a more stable interface between SDK and exporters will be needed or a customer provided exporter DLL can be broken easily with SDK DLL upgrade. So we prefer to build SDK and exporters in 1 DLL. But this can be change if there is valid request.

The single DLL includes SDK and selected exporter which the customer can call into. If the user needs an exporter which is not in the existing DLL, they need build it as static or request the DLL owner to add it.
> If somehow a single DDL is desirable, please describe how is it supposed to be used, I just don't get it.

Reply:
With customizing std::default_delete removed, this was updated as a warning message.